src/ripple/overlay/Slot.h

```diff
@@ -0,0 +1,635 @@
+std::unordered_map<
+    typename Peer::id_t,
+    std::tuple<PeerState, uint16_t, uint32_t, uint32_t>>
+Slot<clock_type>::getPeers() const
+{
+    auto init = std::unordered_map<
+        id_t,
+        std::tuple<PeerState, std::uint16_t, std::uint32_t, std::uint32_t>>();
+    return std::accumulate(
+        peers_.begin(), peers_.end(), init, [](auto& init, auto const& it) {
+            init.emplace(std::make_pair(
+                it.first,
+                std::move(std::make_tuple(
+                    it.second.state,
+                    it.second.count,
+                    epoch<milliseconds>(it.second.expire).count(),
+                    epoch<milliseconds>(it.second.lastMessage).count()))));
+            return init;
+        });
+}
+
+/** Slots is a container for validator's Slot and handles Slot update
+ * when a message is received from a validator. It also handles Slot aging
+ * and checks for peers which are disconnected or stopped relaying the messages.
+ */
+template <typename clock_type>
+class Slots final
+{
+    using time_point = typename clock_type::time_point;
+    using id_t = typename Peer::id_t;
+
+public:
+    /**
+     * @param handler Squelch/unsquelch implementation
+     */
+    Slots(SquelchHandler const& handler) : handler_(handler)
+    {
+    }
+    ~Slots() = default;
+    /** Calls Slot::update of Slot associated with the validator.
+     * @param validator Validator's public key
+     * @param id Peer's id which received the message
+     * @param id Peer's pointer which received the message
+     * @param type Received protocol message type
+     */
+    void
+    checkForSquelch(
+        PublicKey const& validator,
+        id_t id,
+        std::weak_ptr<Peer> const& peer,
+        protocol::MessageType type);
+
+    /** Called when a peer is deleted. If the peer is selected to be the
+     * source of messages from the validator then unsquelch handler
+     * is called for all currently squelched peers.
+     * @param id Peer's id
+     */
+    void
+    unsquelch(id_t id);
+
+    /** Check if peers stopped relaying messages
+     * and if slots stopped receiving messages from the validator.
+     */
+    void
+    checkIdle();
+
+    /** Return number of peers in state in state */
+    boost::optional<std::uint16_t>
+    inState(PublicKey const& validator, PeerState state) const
+    {
+        auto const& it = slots_.find(validator);
+        if (it != slots_.end())
+            return it->second.inState(state);
+        return {};
+    }
+
+    /** Return number of peers in state in state */
+    boost::optional<std::uint16_t>
+    notInState(PublicKey const& validator, PeerState state) const
+    {
+        auto const& it = slots_.find(validator);
+        if (it != slots_.end())
+            return it->second.notInState(state);
+        return {};
+    }
+
+    /** Return true if Slot is in State */
+    template <typename Comp = std::equal_to<>>
+    boost::optional<bool>
```

Comment:
`optional<bool>` is highly error prone to use, and the callers don't appear to need to distinguish between the two levels for `false`. Consider changing the return type to just `bool` (or a struct if it really is necessary to distinguish between the two `false`'s).